### PR TITLE
[operator] Manage `istio` deployments

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area dev-productivity ops-productivity delivery
/kind enhancement

**What this PR does / why we need it**:
With this PR, Gardener-Operator deploys and manages `istiod` and `istio-ingress` components, similar to how Gardenlet deploys them into seed clusters. The `kube-apiserver` service is not yet configured to be reachable via Istio. This will be contributed in a separate PR.

**Which issue(s) this PR fixes**:
Part of #7016

**Release note**:

```feature operator
`gardener-operator` now deploys `Istio` components into the garden runtime cluster.
```

### CODE REVIEW

File: pkg/operation/botanist/component/istio/istiod.go [hunk 1]
```diff
@@ -278,9 +278,6 @@ func (i *istiod) generateIstiodChart(ignoreMode bool) (*chartrenderer.RenderedCh
 	})
 }
 
-func resourceName(name, suffix string) string {
-	if len(suffix) == 0 {
-		return name
-	}
-	return fmt.Sprintf("%s-%s", name, suffix)
+func resourceName(name, prefix string) string {
```
Comment (rfranzke) [inline pkg/operation/botanist/component/istio/istiod.go]: TBH, this function seems useless. Can we drop it and just concatenate the arguments where this function is invoked?

File: pkg/operation/botanist/component/shared/istio.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	priorityClassName string,
+	istiodEnabled bool,
+	labels map[string]string,
+	lbAnnotations map[string]string,
+	externalTrafficPolicy *corev1.ServiceExternalTrafficPolicyType,
+	serviceExternalIP *string,
+	servicePorts []corev1.ServicePort,
+	proxyProtocolEnabled bool,
+	vpnEnabled bool,
+	zones []string) (istio.Interface, error) {
```
Comment (rfranzke) [inline pkg/operation/botanist/component/shared/istio.go]:
```suggestion
	zones []string,
) (
	istio.Interface,
	error,
) {
```

File: pkg/operation/botanist/component/shared/istio.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+		// zonal load balancer is exposed individually via its own IP address. Therefore, having
+		// just one replica may negatively affect availability.
+		minReplicas = pointer.Int(len(zones) * 2)
+		// The default configuration without availability zones has 5 as the maximum amount of
+		// replicas, which apparently works in all known Gardener scenarios. Reducing it to less
+		// per zone gives some room for autoscaling while it is assumed to never reach the maximum.
+		maxReplicas = pointer.Int(len(zones) * 4)
+	}
+
+	ingressGatewayNamespace := fmt.Sprintf("%s%s", namePrefix, ingressNamespace)
```
Comment (rfranzke) [inline pkg/operation/botanist/component/shared/istio.go]: You could inline this variable, it's not used elsewhere.

File: pkg/operation/botanist/component/shared/istio.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+// to fill out common chart values. Hence, it is assumed that at least one Ingress Gateway was added to the given
+// `istioDeployer` before calling this function.
+func AddIstioIngressGateway(
+	istioDeployer istio.Interface,
+	namespace string,
+	annotations map[string]string,
+	labels map[string]string,
+	externalTrafficPolicy *corev1.ServiceExternalTrafficPolicyType,
+	serviceExternalIP *string,
+	zone *string) error {
```
Comment (rfranzke) [inline pkg/operation/botanist/component/shared/istio.go]:
```suggestion
	zone *string,
) error {
```

File: pkg/gardenlet/controller/seed/seed/components.go [hunk 5]
```diff
@@ -80,145 +79,100 @@ func defaultIstio(
...
-				VPNEnabled:           true,
-			})
+			if err := shared.AddIstioIngressGateway(
+				istioDeployer,
+				namespace,
+				seed.GetZonalLoadBalancerServiceAnnotations(zone),
+				operation.GetIstioZoneLabels(labels, &zone),
+				seed.GetZonalLoadBalancerServiceExternalTrafficPolicy(zone),
+				nil,
+				&zone); err != nil {
```
Comment (rfranzke) [inline pkg/gardenlet/controller/seed/seed/components.go]:
```suggestion
				&zone,
			); err != nil {
```
```

File: pkg/gardenlet/controller/seed/seed/components.go [hunk 6]
```diff
@@ -80,145 +79,100 @@ func defaultIstio(
...
-			VPNEnabled:            true,
-		})
+		if err := shared.AddIstioIngressGateway(
+			istioDeployer,
+			*handler.SNI.Ingress.Namespace,
+			utils.MergeStringMaps(seed.GetLoadBalancerServiceAnnotations(), handler.LoadBalancerService.Annotations),
+			operation.GetIstioZoneLabels(gardenerutils.GetMandatoryExposureClassHandlerSNILabels(handler.SNI.Ingress.Labels, handler.Name), nil),
+			seed.GetLoadBalancerServiceExternalTrafficPolicy(),
+			handler.SNI.Ingress.ServiceExternalIP,
+			nil); err != nil {
```
Comment (rfranzke) [inline pkg/gardenlet/controller/seed/seed/components.go]:
```suggestion
			nil,
		); err != nil {
```

File: pkg/gardenlet/controller/seed/seed/components.go [hunk 7]
```diff
@@ -80,145 +79,100 @@ func defaultIstio(
...
-					VPNEnabled:           true,
-				})
+				if err := shared.AddIstioIngressGateway(
+					istioDeployer,
+					namespace,
+					utils.MergeStringMaps(handler.LoadBalancerService.Annotations, seed.GetZonalLoadBalancerServiceAnnotations(zone)),
+					operation.GetIstioZoneLabels(gardenerutils.GetMandatoryExposureClassHandlerSNILabels(handler.SNI.Ingress.Labels, handler.Name), &zone),
+					seed.GetZonalLoadBalancerServiceExternalTrafficPolicy(zone),
+					nil,
+					&zone); err != nil {
```
Comment (rfranzke) [inline pkg/gardenlet/controller/seed/seed/components.go]:
```suggestion
					&zone,
				); err != nil {
```

File: test/integration/gardenlet/seed/seed/seed_test.go [hunk 8]
```diff
@@ -462,20 +462,24 @@ var _ = Describe("Seed controller tests", func() {
 							g.Expect(testClient.Status().Patch(ctx, deployment, patch)).To(Succeed())
 						}).Should(Succeed())
 					} else {
-						// Usually, the gardener-operator would deploy gardener-resource-manager and the related CRD for
-						// ManagedResources and VerticalPodAutoscaler. However, it is not really running, so we have to fake its behaviour here.
-						By("Create CustomResourceDefinition for ManagedResources")
+						// Usually, the gardener-operator deploys and managed the following resource.
```
Comment (rfranzke) [inline test/integration/gardenlet/seed/seed/seed_test.go]:
```suggestion
						// Usually, the gardener-operator deploys and manages the following resources.
```

File: test/integration/gardenlet/seed/seed/seed_test.go [hunk 9]
```diff
@@ -462,20 +462,24 @@ var _ = Describe("Seed controller tests", func() {
...
+						// However, it is not really running, so we have to fake its behaviour here.
+						By("Create resources managed by Gardener-Operator")
```
Comment (rfranzke) [inline test/integration/gardenlet/seed/seed/seed_test.go]:
```suggestion
						By("Create resources managed by gardener-operator")
```

File: pkg/operator/controller/garden/components.go [hunk 10]
```diff
@@ -393,3 +395,30 @@ func fetchKubeconfigFromSecret(ctx context.Context, c client.Client, key client.
...
+	var annotations map[string]string
+	if settings := garden.Spec.RuntimeCluster.Settings; settings != nil && settings.LoadBalancerServices != nil {
+		annotations = settings.LoadBalancerServices.Annotations
+	}
+
+	return sharedcomponent.NewIstio(
+		r.RuntimeClientSet.Client(),
+		r.ImageVector,
+		r.RuntimeClientSet.ChartRenderer(),
+		namePrefix,
```
Comment (rfranzke) [inline pkg/operator/controller/garden/components.go:419]: Is it really necessary to use this prefix? Can we avoid it?

File: pkg/operator/controller/garden/components.go [hunk 11]
```diff
@@ -393,3 +395,30 @@ func fetchKubeconfigFromSecret(ctx context.Context, c client.Client, key client.
...
+		v1beta1constants.DefaultSNIIngressNamespace,
+		v1beta1constants.PriorityClassNameGardenSystem500,
```
Comment (rfranzke) [inline pkg/operator/controller/garden/components.go]: Please adapt https://github.com/gardener/gardener/blob/master/docs/development/priority-classes.md#priorityclasses-for-garden-control-plane-components

File: pkg/operator/controller/garden/reconciler_reconcile.go [hunk 12]
```diff
@@ -192,6 +203,11 @@ func (r *Reconciler) reconcile(
 			Fn:           flow.Parallel(etcdMain.Wait, etcdEvents.Wait),
 			Dependencies: flow.NewTaskIDs(deployEtcds),
 		})
+		_ = g.Add(flow.Task{
+			Name:         "Deploying Istio",
+			Fn:           r.deployIstioFunc(istio),
+			Dependencies: flow.NewTaskIDs(syncPointSystemComponents),
+		})
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_reconcile.go]: I would rather consider istio a system component, so let's move it up to the first "block".

File: pkg/operator/controller/garden/reconciler_reconcile.go [hunk 13]
```diff
@@ -425,3 +441,41 @@ func (r *Reconciler) initializeVirtualClusterClient(secretsManager secretsmanage
...
 }
+
+func (r *Reconciler) deployIstioFunc(istioDeploy istio.Interface) flow.TaskFn {
+	return func(ctx context.Context) error {
+		if err := istioDeploy.Deploy(ctx); err != nil {
+			return err
+		}
+
+		// Deploy NetworkPolicy allowing kube-apiserver to talk all targets (e.g. etcd) and to receive traffic from outside.
+		// TODO(timuthy): Remove this in the future when the network policy deployment has been refactored.
```
Comment (rfranzke) [inline pkg/operator/controller/garden/reconciler_reconcile.go]:
```suggestion
		// TODO(rfranzke): Remove this in the future when the network policy deployment has been refactored.
```

File: pkg/utils/gardener/seed_test.go [hunk 14]
```diff
@@ -175,4 +178,44 @@ var _ = Describe("utils", func() {
 			Expect(err).NotTo(HaveOccurred())
 		})
 	})
+
+	Describe("#SeedIsGarden", func() {
+		var (
+			ctx        context.Context
+			mockClient *mockclient.MockClient
```
Comment (rfranzke) [inline pkg/utils/gardener/seed_test.go:185]: Do not use a mock client

**Review by rfranzke — CHANGES_REQUESTED** [review]
/cc @ScheererJ

File: pkg/gardenlet/controller/seed/seed/components.go [hunk 15]
```diff
@@ -80,145 +79,100 @@ func defaultIstio(
...
+	)
+	if err != nil {
+		return nil, err
 	}
 
 	// Automatically create ingress gateways for single-zone control planes on multi-zonal seeds
 	if len(seedObj.Spec.Provider.Zones) > 1 {
-		for _, zone := range seedObj.Spec.Provider.Zones {
+		for _, z := range seedObj.Spec.Provider.Zones {
+			zone := z
```
Comment (ScheererJ) [inline pkg/gardenlet/controller/seed/seed/components.go]: Why was this change necessary? It seems like zone is only used directly and no pointer to it is stored beyond the end of the loop iteration. Hence, the old approach should work.

File: pkg/gardenlet/controller/seed/seed/components.go [hunk 16]
```diff
@@ -80,145 +79,100 @@ func defaultIstio(
...
+			handler.SNI.Ingress.ServiceExternalIP,
+			nil); err != nil {
+			return nil, err
+		}
 
 		// Automatically create ingress gateways for single-zone control planes on multi-zonal seeds
 		if len(seedObj.Spec.Provider.Zones) > 1 {
-			for _, zone := range seedObj.Spec.Provider.Zones {
+			for _, z := range seedObj.Spec.Provider.Zones {
+				zone := z
```
Comment (ScheererJ) [inline pkg/gardenlet/controller/seed/seed/components.go]: Same as above.

File: pkg/operation/botanist/component/istio/istiod.go [hunk 17]
```diff
@@ -92,13 +94,24 @@ type Values struct {
...
+// Interface contains functions for an Istio deployer.
+type Interface interface {
+	component.DeployWaiter
+	// AddIngressGateway adds another ingress gateway to the existing Istio deployer.
+	AddIngressGateway(values IngressGatewayValues)
+	// GetValues returns the configured values of the Istio deployer.
+	GetValues() Values
+}
+
+var _ Interface = (*istiod)(nil)
```
Comment (ScheererJ) [inline pkg/operation/botanist/component/istio/istiod.go:105]: Why is this needed?

File: pkg/utils/gardener/seed_test.go (hunk 14 above)
Comment (timuthy) [reply pkg/utils/gardener/seed_test.go:185]: Why shouldn't I use a mock client? I actually need it to mock the `NoResourceMatchError` error. I'm not aware of any way to do the same with the fake client.

File: pkg/operator/controller/garden/components.go (hunk 10 above)
Comment (timuthy) [reply pkg/operator/controller/garden/components.go:419]: We need some kind of prefix or suffix in order to name the managed resources and the istio-ingress namespace accordingly. Otherwise we will have a name overlap if the cluster is registered as a seed, as well.

File: pkg/gardenlet/controller/seed/seed/components.go (hunk 15 above)
Comment (timuthy) [reply pkg/gardenlet/controller/seed/seed/components.go]: > It seems like zone is only used directly and no pointer to it is stored beyond the end of the loop iteration

A pointer to the zone is used when `shared.AddIstioIngressGateway` is called. Am I missing something?

File: pkg/operation/botanist/component/istio/istiod.go (hunk 17 above)
Comment (timuthy) [reply pkg/operation/botanist/component/istio/istiod.go:105]: It's a helper to get compiler errors in case `istiod` doesn't implement the `Interface` interface. We use this technique at various places, e.g. [here](https://github.com/gardener/gardener/blob/0b5cdc002b481e77e94d6fae2a1dc9daf1a3a550/pkg/utils/secrets/manager/manager.go#L112).

File: pkg/gardenlet/controller/seed/seed/components.go (hunk 15 above)
Comment (ScheererJ) [reply pkg/gardenlet/controller/seed/seed/components.go]: Yes, but the pointer is dereferenced within `shared.AddIstioIngressGateway` and not stored. Therefore, it should be fine as long as the loop is not executed concurrently, which is not the case as far as I know.

It is not wrong to do it as is. I was just confused as it seems unnecessary at the moment.

Comment (rfranzke) [reply pkg/gardenlet/controller/seed/seed/components.go]: We could also use `pointer.String()` to get rid of these lines?

File: pkg/operator/controller/garden/components.go (hunk 10 above)
Comment (rfranzke) [reply pkg/operator/controller/garden/components.go:419]: OK, but this only concerns the names of the global resources (MR + namespace), or will the prefix be used for more than that?

Comment (timuthy) [reply pkg/operator/controller/garden/components.go:419]: Yes, it's only related to resource names.

Comment (rfranzke) [reply pkg/operator/controller/garden/components.go:419]: Only the names of `ManagedResource` + `Namespace` or also other things (`Deployment`, ...)?

Comment (timuthy) [reply pkg/operator/controller/garden/components.go:419]: For the following resources:
- Ingress namespace for virtual Garden, see [here](https://github.com/gardener/gardener/blob/fcab85c54a0248c273de3a9ad48c00899458c855/pkg/operation/botanist/component/shared/istio.go#L78)
- ManagedResource names, see [here](https://github.com/gardener/gardener/blob/fcab85c54a0248c273de3a9ad48c00899458c855/pkg/operation/botanist/component/istio/istiod.go#L120-L121)

Comment (rfranzke) [reply pkg/operator/controller/garden/components.go:419]: 😀 Great, OK, thank you!

File: docs/development/priority-classes.md [hunk 18]
```diff
@@ -19,14 +19,14 @@ When using the `gardener-operator` for managing the garden runtime and virtual c
...
-| `gardener-garden-system-400`      | 999999400 |                                                                                           |
-| `gardener-garden-system-300`      | 999999300 | `vpa-admission-controller`, `etcd-druid`                                                  |
-| `gardener-garden-system-200`      | 999999200 | `vpa-recommender`, `vpa-updater`, `hvpa-controller`                                       |
-| `gardener-garden-system-100`      | 999999100 |                                                                                           |
+| Name                              | Priority  | Associated Components (Examples)                                                                    |
+|---------------------------------- |-----------|-----------------------------------------------------------------------------------------------------|
+| `gardener-garden-system-critical` | 999999550 | `gardener-operator`, `gardener-resource-manager`                                                    |
+| `gardener-garden-system-500`      | 999999500 | `virtual-garden-etcd-events`, `virtual-garden-etcd-main`, `virtual-garden-kube-apiserver`, `istio`  |
```
Comment (rfranzke) [inline docs/development/priority-classes.md]: In seeds, istio does have the same priority like gardenlet/GRM (ref https://github.com/gardener/gardener/blob/878377d9cb41214634fd01b48c035da190304eba/docs/development/priority-classes.md#priorityclasses-for-seed-system-components). Perhaps we should do the same in garden clusters, i.e. assign `gardener-garden-system-critical`?

File: pkg/gardenlet/controller/seed/seed/components.go [hunk 19]
```diff
@@ -76,145 +74,101 @@ func defaultIstio(
...
-				Zones:                []string{zone},
-				Namespace:            namespace,
-				ProxyProtocolEnabled: features.DefaultFeatureGate.Enabled(features.APIServerSNI),
-				VPNEnabled:           true,
-			})
+			namespace := shared.GetIstioNamespaceForZone(*conf.SNI.Ingress.Namespace, zone)
+
+			if err := shared.AddIstioIngressGateway(
+				istioDeployer,
+				namespace,
```
Comment (rfranzke) [inline pkg/gardenlet/controller/seed/seed/components.go]:
```suggestion
			if err := shared.AddIstioIngressGateway(
				istioDeployer,
				shared.GetIstioNamespaceForZone(*conf.SNI.Ingress.Namespace, zone),
```

File: pkg/gardenlet/controller/seed/seed/components.go [hunk 20]
```diff
@@ -76,145 +74,101 @@ func defaultIstio(
...
-					Zones:                []string{zone},
-					Namespace:            namespace,
-					ProxyProtocolEnabled: features.DefaultFeatureGate.Enabled(features.APIServerSNI),
-					VPNEnabled:           true,
-				})
+				namespace := shared.GetIstioNamespaceForZone(*handler.SNI.Ingress.Namespace, zone)
+
+				if err := shared.AddIstioIngressGateway(
+					istioDeployer,
+					namespace,
```
Comment (rfranzke) [inline pkg/gardenlet/controller/seed/seed/components.go]:
```suggestion
				if err := shared.AddIstioIngressGateway(
					istioDeployer,
					shared.GetIstioNamespaceForZone(*handler.SNI.Ingress.Namespace, zone),
```

**Review by ScheererJ — COMMENTED** [review]
/lgtm

**Review by rfranzke — COMMENTED** [review]
Could you extend `test/integration/gardenlet/seed/seed/seed_test.go` to also check for the expected Istio-related `ManagedResource`s in the `istio-system` namespace?

**Comment by timuthy** [comment]
Rebased PR and enhanced integration test as requested by https://github.com/gardener/gardener/pull/7817#pullrequestreview-1403356771.

**Review by rfranzke — COMMENTED** [review]
Nice, let's go! :)

/lgtm
/approve
